lib/web/app/middleware.go

```diff
@@ -77,16 +78,19 @@ func (h *Handler) redirectToLauncher(w http.ResponseWriter, r *http.Request, p l
 	if p.clusterName != "" && p.publicAddr != "" {
 		urlPath = append(urlPath, p.clusterName, p.publicAddr)
 	}
-	urlQuery := ""
+	var urlQuery []string
 	if p.stateToken != "" {
-		urlQuery = fmt.Sprintf("state=%v", p.stateToken)
+		urlQuery = append(urlQuery, fmt.Sprintf("state=%v", p.stateToken))
+	}
+	if p.awsRole != "" {
+		urlQuery = append(urlQuery, fmt.Sprintf("awsrole=%v", p.awsRole))
 	}
 
 	u := url.URL{
 		Scheme:   "https",
 		Host:     h.c.WebPublicAddr,
 		Path:     path.Join(urlPath...),
-		RawQuery: urlQuery,
+		RawQuery: strings.Join(urlQuery, "&"),
```

Comment:
Should these query parameters be escaped, or do we always expect `p.stateToken` and `p.awsRole` to have sane content?

Reply:
Yeah it makes sense to escape role ARN since it contains special characters.

@rudream When I try to escape it on the backend though it stops working because the slash in the ARN gets double-escaped at some point so what I get on the backend (in /x-teleport-auth handler, after Go unescapes the query string) is `arn:aws:iam::1234:role%2FEC2ReadOnly` - note `%2F` instead of `/`.

I suspect this happens because initially frontend encodes this role ARN as a part of path with `/` escaped and maybe it doesn't unescape it before inserting into the query string afterwards? Could you make sure that role ARN that gets sent to the backend as a part of the query string is escaped properly? Then I can do proper escaping here as well.